easy_perception_deployment/include/ort_cpp_lib/p3_ort_base.cpp

```diff
@@ -956,15 +1299,559 @@ EPD::EPDObjectLocalization P3OrtBase::infer_action(
+        cv::Point(curBbox[2], curBbox[3])) / 2;
+
+      obj_surface_depth = this->findMin(depthImg(curBoxRect)) * 0.001;
+      float x = (rotated_mid.x - ppx) / fx * obj_surface_depth;
+      float y = (rotated_mid.y - ppy) / fy * obj_surface_depth;
+
+      output_obj.objects[i].centroid.x = x;
+      output_obj.objects[i].centroid.y = y;
+      output_obj.objects[i].centroid.z = obj_surface_depth +
+        (table_depth - obj_surface_depth) / 2;
+
+      // Get Real Size and angle of object
+      // Compare the length of 2 side of the rectangle,
+      // the longer side will be the major axis
+      if (cv::norm(rect_points[0] - rect_points[1]) >
+        cv::norm(rect_points[1] - rect_points[2]))
+      {
+        // Calculates the length of the object
+        output_obj.objects[i].length = obj_surface_depth * sqrt(
+          pow((pt_a.x - pt_b.x) / fx, 2) +
+          pow((pt_a.y - pt_b.y) / fy, 2));
+        // Calculates the breadth of the object
+        output_obj.objects[i].breadth = obj_surface_depth * sqrt(
+          pow((pt_c.x - pt_d.x) / fx, 2) +
+          pow((pt_c.y - pt_d.y) / fy, 2));
+      } else {
+        // Gets object breadth and length
+        output_obj.objects[i].breadth = obj_surface_depth * sqrt(
+          pow((pt_a.x - pt_b.x) / fx, 2) +
+          pow((pt_a.y - pt_b.y) / fy, 2));
+        output_obj.objects[i].length = obj_surface_depth * sqrt(
+          pow((pt_c.x - pt_d.x) / fx, 2) +
+          pow((pt_c.y - pt_d.y) / fy, 2));
+      }
+      // Setting height of object
+      output_obj.objects[i].height = table_depth - obj_surface_depth;
+
+      pcl::PointCloud<pcl::PointXYZ>::Ptr segmented_cloud(new pcl::PointCloud<pcl::PointXYZ>);
+      segmented_cloud->header.frame_id = "camera_color_optical_frame";
+      segmented_cloud->is_dense = true;
+
+      cv::Mat tempImg = inputImg.clone();
+
+      // Converting Depth Image to PointCloud
+      for (int j = 0; j < tempFinalMask.rows; j++) {
+        for (int k = 0; k < tempFinalMask.cols; k++) {
+          // TODO(cardboardcode) convert segmented mask into segmented pointcloud
+          int pixelValue = static_cast<int>(tempFinalMask.at<uchar>(j, k));
+
+          if (pixelValue != 0) {
+            float z = static_cast<float>(depthImg.at<uint16_t>(
+                curBoxRect.y + j, curBoxRect.x + k) * 0.001);
+            float x = static_cast<float>((curBoxRect.x + k - ppx) / fx) * z;
+            float y = static_cast<float>((curBoxRect.y + j - ppy) / fy) * z;
+
+            // Ignore all points that has a value of less than 0.1mm in z.
+            if (std::abs(z) < 0.0001 || std::abs(z) > camera_to_plane_distance_mm * 0.001) {
+              continue;
+            } else {
+              pcl::PointXYZ curPoint(x, y, z);
+              segmented_cloud->points.push_back(curPoint);
+            }
+          }
+        }
+      }
+
+      output_obj.objects[i].segmented_pcl = *segmented_cloud;
```

Comment:
Yes. It would be more efficient but not significantly high. For debugging reasons, I have left it as such because I foresee more optimization done in this specific part of the Localization feature in the near future. 

This trade-off emphasizes readability of code, at least for me as EPD's developer.

If more optimization is required for Localization, I will modify the code to add information directly into `output_obj.objects[i].segmented_pcl`. Will leave it unchanged for now.